File: GdPicture/WinForm/vb.net/Segmentation/Utilities.vb
Imports GdPicture14

Friend Class Utilities

    Shared ReadOnly BlockColor As Color = Color.Green
    Shared ReadOnly ParagraphColor As Color = Color.Cyan
    Shared ReadOnly TextLineColor As Color = Color.Magenta
    Shared ReadOnly WordColor As Color = Color.Blue
    Shared ReadOnly CharacterColor As Color = Color.Red

    Public Shared Sub MarkRegions(ByVal gdPictureSegmenter As GdPictureSegmenter, ByVal resultID As String, ByVal gdViewer As GdViewer, ByVal showBlocks As Boolean,
                                  ByVal showParagraphs As Boolean, ByVal showTextLines As Boolean, ByVal showWords As Boolean, ByVal showCharacters As Boolean,
                                  ByVal showText As Boolean, ByVal showBitmap As Boolean, ByVal showLines As Boolean, ByVal horizontalResolution As Single, ByVal verticalResolution As Single)
        gdViewer.RemoveAllRegions()

        Dim blocks As List(Of Rectangle)
        Dim paragraphs As List(Of Rectangle)
        Dim lines As List(Of Rectangle)
        Dim words As List(Of Rectangle)
        Dim chars As List(Of Rectangle)

        GetElementsBoundingBoxes(gdPictureSegmenter, resultID, showText, showBitmap, showLines, blocks, paragraphs, lines, words, chars)
        If showBlocks Then
            AddRegions(gdViewer, blocks, BlockColor, horizontalResolution, verticalResolution)
        End If

        If showParagraphs Then
            AddRegions(gdViewer, paragraphs, ParagraphColor, horizontalResolution, verticalResolution)
        End If

        If showTextLines Then
            AddRegions(gdViewer, lines, TextLineColor, horizontalResolution, verticalResolution)
        End If

        If showWords Then
            AddRegions(gdViewer, words, WordColor, horizontalResolution, verticalResolution)
        End If

        If showCharacters Then
            AddRegions(gdViewer, chars, CharacterColor, horizontalResolution, verticalResolution)
        End If

    End Sub

    Private Shared Sub GetElementsBoundingBoxes(ByVal gdPictureSegmenter As GdPictureSegmenter, ByVal segmentationResultId As String, ByVal getText As Boolean,
                                               ByVal getBitmap As Boolean, ByVal getLines As Boolean, ByRef blocks As List(Of Rectangle), ByRef paragraphs As List(Of Rectangle),
                                               ByRef lines As List(Of Rectangle), ByRef words As List(Of Rectangle), ByRef chars As List(Of Rectangle))
        blocks = New List(Of Rectangle)
        paragraphs = New List(Of Rectangle)
        lines = New List(Of Rectangle)
        words = New List(Of Rectangle)
        chars = New List(Of Rectangle)
        Dim blocksCount As Integer = gdPictureSegmenter.GetBlockCount(segmentationResultId)
        Dim blockIndex As Integer = 0
        Do While blockIndex < blocksCount
            Dim blockType As OCRBlockType = gdPictureSegmenter.GetBlockType(segmentationResultId, blockIndex)
            If (getText AndAlso IsTextBlockType(blockType)) OrElse (getBitmap AndAlso IsImageBlockType(blockType)) OrElse (getLines AndAlso IsLineBlockType(blockType)) Then
                blocks.Add(New Rectangle(gdPictureSegmenter.GetBlockLeft(segmentationResultId, blockIndex), gdPictureSegmenter.GetBlockTop(segmentationResultId, blockIndex), gdPictureSegmenter.GetBlockRight(segmentationResultId, blockIndex) - gdPictureSegmenter.GetBlockLeft(segmentationResultId, blockIndex), gdPictureSegmenter.GetBlockBottom(segmentationResultId, blockIndex) - gdPictureSegmenter.GetBlockTop(segmentationResultId, blockIndex)))
                Dim paragraphsCount As Integer = gdPictureSegmenter.GetBlockParagraphCount(segmentationResultId, blockIndex)
                Dim paragraphIndex As Integer = gdPictureSegmenter.GetBlockFirstParagraphIndex(segmentationResultId, blockIndex)
                Do While paragraphIndex < (gdPictureSegmenter.GetBlockFirstParagraphIndex(segmentationResultId, blockIndex) + paragraphsCount)
                    paragraphs.Add(New Rectangle(gdPictureSegmenter.GetParagraphLeft(segmentationResultId, paragraphIndex), gdPictureSegmenter.GetParagraphTop(segmentationResultId, paragraphIndex), gdPictureSegmenter.GetParagraphRight(segmentationResultId, paragraphIndex) - gdPictureSegmenter.GetParagraphLeft(segmentationResultId, paragraphIndex), gdPictureSegmenter.GetParagraphBottom(segmentationResultId, paragraphIndex) - gdPictureSegmenter.GetParagraphTop(segmentationResultId, paragraphIndex)))
                    Dim textLinesCount As Integer = gdPictureSegmenter.GetParagraphTextLineCount(segmentationResultId, paragraphIndex)
                    Dim textLineIndex As Integer = gdPictureSegmenter.GetParagraphFirstTextLineIndex(segmentationResultId, paragraphIndex)
                    Do While textLineIndex < (gdPictureSegmenter.GetParagraphFirstTextLineIndex(segmentationResultId, paragraphIndex) + textLinesCount)
                        lines.Add(New Rectangle(gdPictureSegmenter.GetTextLineLeft(segmentationResultId, textLineIndex), gdPictureSegmenter.GetTextLineTop(segmentationResultId, textLineIndex), gdPictureSegmenter.GetTextLineRight(segmentationResultId, textLineIndex) - gdPictureSegmenter.GetTextLineLeft(segmentationResultId, textLineIndex), gdPictureSegmenter.GetTextLineBottom(segmentationResultId, textLineIndex) - gdPictureSegmenter.GetTextLineTop(segmentationResultId, textLineIndex)))
                        Dim wordsCount As Integer = gdPictureSegmenter.GetTextLineWordCount(segmentationResultId, textLineIndex)
                        Dim wordIndex As Integer = gdPictureSegmenter.GetTextLineFirstWordIndex(segmentationResultId, textLineIndex)
                        Do While wordIndex < (gdPictureSegmenter.GetTextLineFirstWordIndex(segmentationResultId, textLineIndex) + wordsCount)
                            words.Add(New Rectangle(gdPictureSegmenter.GetWordLeft(segmentationResultId, wordIndex), gdPictureSegmenter.GetWordTop(segmentationResultId, wordIndex), gdPictureSegmenter.GetWordRight(segmentationResultId, wordIndex) - gdPictureSegmenter.GetWordLeft(segmentationResultId, wordIndex), gdPictureSegmenter.GetWordBottom(segmentationResultId, wordIndex) - gdPictureSegmenter.GetWordTop(segmentationResultId, wordIndex)))
                            Dim charactersCount As Integer = gdPictureSegmenter.GetWordCharacterCount(segmentationResultId, wordIndex)
                            Dim characterIndex As Integer = gdPictureSegmenter.GetWordFirstCharacterIndex(segmentationResultId, wordIndex)
                            Do While characterIndex < (gdPictureSegmenter.GetWordFirstCharacterIndex(segmentationResultId, wordIndex) + charactersCount)
                                chars.Add(New Rectangle(gdPictureSegmenter.GetCharacterLeft(segmentationResultId, characterIndex), gdPictureSegmenter.GetCharacterTop(segmentationResultId, characterIndex), gdPictureSegmenter.GetCharacterRight(segmentationResultId, characterIndex) - gdPictureSegmenter.GetCharacterLeft(segmentationResultId, characterIndex), gdPictureSegmenter.GetCharacterBottom(segmentationResultId, characterIndex) - gdPictureSegmenter.GetCharacterTop(segmentationResultId, characterIndex)))
                                characterIndex = characterIndex + 1
                            Loop

                            wordIndex = wordIndex + 1
                        Loop

                        textLineIndex = textLineIndex + 1
                    Loop

                    paragraphIndex = paragraphIndex + 1
                Loop

            End If

            blockIndex = blockIndex + 1
        Loop

    End Sub

    Private Shared Function IsImageBlockType(ByVal blockType As OCRBlockType) As Boolean
        Return (blockType = OCRBlockType.FlowingImage) OrElse (blockType = OCRBlockType.HeadingImage) OrElse (blockType = OCRBlockType.PulloutImage)
    End Function

    Private Shared Function IsTextBlockType(ByVal blockType As OCRBlockType) As Boolean
        Return (blockType = OCRBlockType.CaptionText) OrElse (blockType = OCRBlockType.FlowingText) OrElse (blockType = OCRBlockType.HeadingText) OrElse (blockType = OCRBlockType.PulloutText) OrElse (blockType = OCRBlockType.Table)
    End Function

    Private Shared Function IsLineBlockType(ByVal blockType As OCRBlockType) As Boolean
        Return (blockType = OCRBlockType.VerticalLine) OrElse (blockType = OCRBlockType.HorizontalLine)
    End Function

    Private Shared Sub AddRegions(ByVal gdViewer As GdViewer, ByVal rectangles As IEnumerable(Of Rectangle),
                                  ByVal color As Color, ByVal horizontalResolution As Single, ByVal verticalResolution As Single)
        For Each rectangle In rectangles
            Dim regionId As Integer = gdViewer.AddRegionInches("", rectangle.Left / horizontalResolution, rectangle.Top / verticalResolution, (rectangle.Right - rectangle.Left) / horizontalResolution, (rectangle.Bottom - rectangle.Top) / verticalResolution, ForegroundMixMode.ForegroundMixModeUNDEFINED, color)
            gdViewer.SetRegionEditable(regionId, False)
            gdViewer.SetRegionBorderWidth(regionId, 1)
            gdViewer.SetRegionBorderColor(regionId, color)
        Next
    End Sub

End Class

File: GdPicture/WinForm/C#/Segmentation/Utilities.cs
using GdPicture14;
using System.Collections.Generic;
using System.Drawing;


namespace Segmenter
{
    static class Utilities
    {
        /// <summary>
        /// The color to be used for blocks.
        /// </summary>
        private static readonly Color BlockColor = Color.Green;

        /// <summary>
        /// The color to be used for paragraphs.
        /// </summary>
        private static readonly Color ParagraphColor = Color.Cyan;

        /// <summary>
        /// The color to be used for text lines.
        /// </summary>
        private static readonly Color TextLineColor = Color.Magenta;

        /// <summary>
        /// The color to be used for words.
        /// </summary>
        private static readonly Color WordColor = Color.Blue;

        /// <summary>
        /// The color to be used for characters.
        /// </summary>
        private static readonly Color CharacterColor = Color.Red;

        public static void MarkRegions(GdPictureSegmenter gdPictureSegmenter, string resultID,
            GdViewer gdViewer, bool showBlocks, bool showParagraphs,
            bool showTextLines, bool showWords, bool showCharacters, 
            bool showText, bool showBitmap, bool showLines,
            float horizontalResolution, float verticalResolution)
        {
            gdViewer.RemoveAllRegions();

            GetElementsBoundingBoxes(gdPictureSegmenter, resultID,
                                     showText, showBitmap, showLines,
                                     out List<Rectangle> blocks,
                                     out List<Rectangle> paragraphs,
                                     out List<Rectangle> lines,
                                     out List<Rectangle> words,
                                     out List<Rectangle> chars);

            if (showBlocks)
            {
                AddRegions(gdViewer, blocks, BlockColor, horizontalResolution, verticalResolution);
            }

            if (showParagraphs)
            {
                AddRegions(gdViewer, paragraphs, ParagraphColor, horizontalResolution, verticalResolution);
            }

            if (showTextLines)
            {
                AddRegions(gdViewer, lines, TextLineColor, horizontalResolution, verticalResolution);
            }

            if (showWords)
            {
                AddRegions(gdViewer, words, WordColor, horizontalResolution, verticalResolution);
            }

            if (showCharacters)
            {
                AddRegions(gdViewer, chars, CharacterColor, horizontalResolution, verticalResolution);
            }
        }


        public static void GetElementsBoundingBoxes(GdPictureSegmenter gdPictureSegmenter,
                                                    string segmentationResultId,
                                                    bool getText, bool getBitmap, bool getLines,
                                                    out List<Rectangle> blocks,
                                                    out List<Rectangle> paragraphs,
                                                    out List<Rectangle> lines,
                                                    out List<Rectangle> words,
                                                    out List<Rectangle> chars)
        {
            blocks = new List<Rectangle>();
            paragraphs = new List<Rectangle>();
            lines = new List<Rectangle>();
            words = new List<Rectangle>();
            chars = new List<Rectangle>();

            int blocksCount = gdPictureSegmenter.GetBlockCount(segmentationResultId);

            for (int blockIndex = 0; blockIndex < blocksCount; blockIndex++)
            {
                OCRBlockType blockType = gdPictureSegmenter.GetBlockType(segmentationResultId, blockIndex);

                if( (blockType == OCRBlockType.Unknown /*type returned by connected components*/) ||
                    (getText && IsTextBlockType(blockType)) ||
                    (getBitmap && IsImageBlockType(blockType)) ||
                    (getLines && IsLineBlockType(blockType))
                    )
                { 

                blocks.Add(new Rectangle(gdPictureSegmenter.GetBlockLeft(segmentationResultId, blockIndex),
                                         gdPictureSegmenter.GetBlockTop(segmentationResultId, blockIndex),
                                         gdPictureSegmenter.GetBlockRight(segmentationResultId, blockIndex) -
                                         gdPictureSegmenter.GetBlockLeft(segmentationResultId, blockIndex),
                                         gdPictureSegmenter.GetBlockBottom(segmentationResultId, blockIndex) -
                                         gdPictureSegmenter.GetBlockTop(segmentationResultId, blockIndex)));
                int paragaphsCount = gdPictureSegmenter.GetBlockParagraphCount(segmentationResultId, blockIndex);

                    for (int paragaphIndex = gdPictureSegmenter.GetBlockFirstParagraphIndex(segmentationResultId, blockIndex); paragaphIndex < gdPictureSegmenter.GetBlockFirstParagraphIndex(segmentationResultId, blockIndex) + paragaphsCount; paragaphIndex++)
                    {
                        paragraphs.Add(new Rectangle(gdPictureSegmenter.GetParagraphLeft(segmentationResultId, paragaphIndex),
                                       gdPictureSegmenter.GetParagraphTop(segmentationResultId, paragaphIndex),
                                       gdPictureSegmenter.GetParagraphRight(segmentationResultId, paragaphIndex) -
                                       gdPictureSegmenter.GetParagraphLeft(segmentationResultId, paragaphIndex),
                                       gdPictureSegmenter.GetParagraphBottom(segmentationResultId, paragaphIndex) -
                                       gdPictureSegmenter.GetParagraphTop(segmentationResultId, paragaphIndex)));
                        int textLinesCount = gdPictureSegmenter.GetParagraphTextLineCount(segmentationResultId, paragaphIndex);

                        for (int textLineIndex = gdPictureSegmenter.GetParagraphFirstTextLineIndex(segmentationResultId, paragaphIndex); textLineIndex < gdPictureSegmenter.GetParagraphFirstTextLineIndex(segmentationResultId, paragaphIndex) + textLinesCount; textLineIndex++)
                        {
                            lines.Add(new Rectangle(gdPictureSegmenter.GetTextLineLeft(segmentationResultId, textLineIndex),
                                                    gdPictureSegmenter.GetTextLineTop(segmentationResultId, textLineIndex),
                                                    gdPictureSegmenter.GetTextLineRight(segmentationResultId, textLineIndex) -
                                                    gdPictureSegmenter.GetTextLineLeft(segmentationResultId, textLineIndex),
                                                    gdPictureSegmenter.GetTextLineBottom(segmentationResultId, textLineIndex) -
                                                    gdPictureSegmenter.GetTextLineTop(segmentationResultId, textLineIndex)));

                            int wordsCount = gdPictureSegmenter.GetTextLineWordCount(segmentationResultId, textLineIndex);
                            for (int wordIndex = gdPictureSegmenter.GetTextLineFirstWordIndex(segmentationResultId, textLineIndex); wordIndex < gdPictureSegmenter.GetTextLineFirstWordIndex(segmentationResultId, textLineIndex) + wordsCount; wordIndex++)
                            {
                                words.Add(new Rectangle(gdPictureSegmenter.GetWordLeft(segmentationResultId, wordIndex),
                                                        gdPictureSegmenter.GetWordTop(segmentationResultId, wordIndex),
                                                        gdPictureSegmenter.GetWordRight(segmentationResultId, wordIndex) -
                                                        gdPictureSegmenter.GetWordLeft(segmentationResultId, wordIndex),
                                                        gdPictureSegmenter.GetWordBottom(segmentationResultId, wordIndex) -
                                                        gdPictureSegmenter.GetWordTop(segmentationResultId, wordIndex)));

                                int charactersCount = gdPictureSegmenter.GetWordCharacterCount(segmentationResultId, wordIndex);
                                for (int characterIndex = gdPictureSegmenter.GetWordFirstCharacterIndex(segmentationResultId, wordIndex); characterIndex < gdPictureSegmenter.GetWordFirstCharacterIndex(segmentationResultId, wordIndex) + charactersCount; characterIndex++)
                                {
                                    chars.Add(new Rectangle(
                                        gdPictureSegmenter.GetCharacterLeft(segmentationResultId, characterIndex),
                                        gdPictureSegmenter.GetCharacterTop(segmentationResultId, characterIndex),
                                        gdPictureSegmenter.GetCharacterRight(segmentationResultId, characterIndex) -
                                        gdPictureSegmenter.GetCharacterLeft(segmentationResultId, characterIndex),
                                        gdPictureSegmenter.GetCharacterBottom(segmentationResultId, characterIndex) -
                                        gdPictureSegmenter.GetCharacterTop(segmentationResultId, characterIndex)));
                                }
                            }
                        }
                    }
                }
            }
        }


        public static bool IsImageBlockType(OCRBlockType blockType)
        {
            return blockType == OCRBlockType.FlowingImage ||
                   blockType == OCRBlockType.HeadingImage ||
                   blockType == OCRBlockType.PulloutImage;
        }


        public static bool IsTextBlockType(OCRBlockType blockType)
        {
            return blockType == OCRBlockType.CaptionText ||
                   blockType == OCRBlockType.FlowingText ||
                   blockType == OCRBlockType.HeadingText ||
                   blockType == OCRBlockType.PulloutText ||
                   blockType == OCRBlockType.Table;
        }


        public static bool IsLineBlockType(OCRBlockType blockType)
        {
            return blockType == OCRBlockType.VerticalLine ||
                   blockType == OCRBlockType.HorizontalLine;
        }


        private static void AddRegions(GdViewer gdViewer, IEnumerable<Rectangle> rectangles, Color color,
            float horizontalResolution, float verticalResolution)
        {
            foreach (var rectangle in rectangles)
            {
                int regionId = gdViewer.AddRegionInches("", rectangle.Left / horizontalResolution,
                    rectangle.Top / verticalResolution, (rectangle.Right - rectangle.Left) / horizontalResolution,
                    (rectangle.Bottom - rectangle.Top) / verticalResolution, ForegroundMixMode.ForegroundMixModeUNDEFINED,
                    color);
                gdViewer.SetRegionEditable(regionId, false);
                gdViewer.SetRegionBorderWidth(regionId, 1);
                gdViewer.SetRegionBorderColor(regionId, color);
            }
        }
    }
}